### PR TITLE
[FEAT] (aptable): add AptaBLE module with symmetric bidirectional cross-attention interaction map

### PR DESCRIPTION
**Pull request by DZDasherKTB**
## Research Background

**Paper:** AptaBLE: Aptamer Binding and Likelihood Estimation
**Venue:** NeurIPS 2024 workshop, updated January 2026
**Key finding:** Symmetric bidirectional cross-attention between aptamer and protein encoder outputs produces interaction maps that recapitulate experimental binding interfaces, outperforming AptaTrans on both DNA and RNA aptamer datasets.

Specifically, AptaBLE investigated SARS-CoV-2 Spike glycoprotein binding aptamers and showed the cross-attention weights identify contacts at three separate binding interfaces, all slightly more active than neighbouring base regions. This means the attention weights are not just a training signal, they are scientifically interpretable contact predictions.

AptaBLE also demonstrated that fine-tuning on as few as two known binder/non-binder pairs for a specific target eliminates false positives that the base AptaTrans model produces, which is the practical motivation for keeping the encoder architecture identical and making pretrained weights transferable.

This PR implements the cross-attention interaction map from AptaBLE as a new `pyaptamer/aptable/` module, structured identically to the existing `pyaptamer/aptatrans/` module for clean integration.

---

## Summary

While studying the AptaTrans architecture, I came across the AptaBLE paper (NeurIPS 2024, updatedJanuary 2026) which directly addresses one of AptaTrans' core limitations.
This PR adds a new `pyaptamer/aptable/` module implementing the AptaBLE architecture as a self-contained extension, structured identically to `pyaptamer/aptatrans/`, with zero changes to any existing code.

---

## The problem with AptaTrans' interaction map

AptaTrans computes the interaction map as a dot product between encoder
outputs:

```python
imap[i, j] = x_apta[i] . x_prot[j]
```

This is a global similarity score. Every aptamer position is scored against every protein position with equal weight. The model has no mechanism to learn that a specific aptamer stem-loop binds a specific protein loop region, it can only detect overall sequence similarity between the two.

---

## What AptaBLE does differently

AptaBLE replaces the dot-product with symmetric bidirectional cross-attention. The aptamer attends to the protein and the protein attends to the aptamer simultaneously:
Aptamer-to-protein: aptamer learns which protein positions matter x_a2p, attn_a2p = CrossAttention(Q=x_apta, K=x_prot, V=x_prot)
Protein-to-aptamer: protein learns which aptamer positions matter x_p2a, attn_p2a = CrossAttention(Q=x_prot, K=x_apta, V=x_apta)
Outer product of attended representations -> interaction map
imap = x_a2p @ x_p2a.T

AptaBLE demonstrated that the resulting attention weight matrices recapitulate experimental binding interfaces on SARS-CoV-2 Spike glycoprotein aptamers, identifying contacts at three separate binding interfaces. The attention weights are directly interpretable as residue-level contact predictions, which resolves the existing TODO in `AptaTransPipeline.get_interaction_map()`.

---

## What this PR adds
pyaptamer/
└── aptable/
├── init.py
├── _model.py                    AptaBLE model
├── _model_lightning.py          Lightning training wrapper
├── layers/
│   ├── init.py
│   └── _cross_attention_map.py  CrossAttentionInteractionMap
└── tests/
├── init.py
└── test_aptable.py          24 unit tests

### `CrossAttentionInteractionMap`

Two `nn.MultiheadAttention` modules running symmetrically. Residual connection and LayerNorm applied before the outer product. Padding masks forwarded from the zero-padded token sequences. Output shape is identical to `InteractionMap`: `(batch, 1, seq_apta, seq_prot)` so the convolutional head requires zero changes.

`return_attention=True` returns the attention weight matrices alongside the interaction map for contact analysis.

### `AptaBLE`

Identical to `AptaTrans` in every respect except `self.imap`. Uses the same encoder architecture, same convolutional head, same fully-connected output. Pretrained AptaTrans encoder weights transfer directly via `load_aptatrans_encoders()`.

### `AptaBLELightning`

Mirrors `AptaTransLightning` interface exactly. Existing training scripts swap `AptaTrans` for `AptaBLE` in one line.

---

## Usage

```python
from pyaptamer.aptable import AptaBLE, AptaBLELightning
from pyaptamer.aptatrans import EncoderPredictorConfig

cfg_a = EncoderPredictorConfig(128, 16, max_len=275)
cfg_p = EncoderPredictorConfig(128, 16, max_len=1000)

model = AptaBLE(cfg_a, cfg_p)

# Transfer pretrained AptaTrans encoders
model.load_aptatrans_encoders(aptatrans_model.state_dict())

# Standard forward pass
preds = model(x_apta, x_prot)

# Interpretable contact maps
imap, attn_a2p, attn_p2a = model.forward_imap(
    x_apta, x_prot, return_attention=True
)
# attn_a2p: (batch, seq_apta, seq_prot), which protein positions each aptamer position attends to
# attn_p2a: (batch, seq_prot, seq_apta), which aptamer positions each protein position attends to
```
---

## Integration note

This module is fully self-contained. No existing files are modified. To expose it at the package level, one line can be added to the root
`pyaptamer/__init__.py`:

```python
from pyaptamer import aptable
```

Alternatively, `CrossAttentionInteractionMap` can be integrated into `AptaTrans` directly as a `use_cross_attention=True` flag. Happy to implement either approach based on maintainer preference.

---

## Note on research and development process

I researched the AptaBLE paper (NeurIPS 2024, January 2026) and studied its architecture in detail to understand where and how it improves on AptaTrans. I used Claude as a research and coding assistant to help navigate the paper and scaffold the implementation. The architecture decisions, the paper references, and the technical correctness of the
implementation have been verified by me personally. 

---

**Author:** Dashpreet Singh | dashpreetsinghhanda@gmail.com
IIT Jammu, B.Tech CSE 2024-2028

### CODE REVIEW

**Comment by DZDasherKTB** [comment]
Hey @satvshr! Wanted to flag this before selections are finalised. This PR adds a standalone pyaptamer/aptable/ module implementing the AptaBLE architecture from NeurIPS 2024, replacing the dot-product InteractionMap in AptaTrans with symmetric bidirectional cross-attention. AptaBLE demonstrated this produces maps that recapitulate experimental binding interfaces on SARS-CoV-2 Spike glycoprotein aptamers. Zero changes to existing code, with clean integration. Would really appreciate your feedback!

**Comment by siddharth7113** [comment]
Hi @DZDasherKTB,

Thank you for implementing this, since this is a large code - dffi , would it be possible for you to add a notebook in `examples` with API to show it is working, this way review for this becomes easier.
